web/packages/teleport/src/Discover/SelectResource/SelectResource.tsx

```diff
@@ -418,6 +411,52 @@ export function sortResources(
   return sortedResources;
 }
 
+/**
+ * Returns prioritized resources based on user preferences cluster state
+ *
+ * @remarks
+ * A user can have preferredResources set via onboarding either from the survey (preferredResources)
+ * or various query parameters (marketingParams). We sort the list by the marketingParam if available.
+ * If not, we sort by preferred resource type if available.
+ * We do not search.
+ *
+ * @param preferences - Cluster state user preferences
+ * @returns PrioritizedResources which is both the resource to prioritize and a boolean value of the value
+ *
+ */
+function getPrioritizedResources(
+  preferences: UserPreferences
+): PrioritizedResources {
+  const marketingParams =
+    preferences && preferences.onboard && preferences.onboard.marketingParams;
+
+  if (marketingParams) {
+    const marketingPriorities = getMarketingMatches(marketingParams);
+    if (marketingPriorities.length > 0) {
+      return {
+        hasPreferredResources: true,
+        preferredResources: marketingPriorities,
+      };
+    }
+  }
+
+  const preferredResources =
+    (preferences &&
+      preferences.onboard &&
+      preferences.onboard.preferredResources) ||
+    [];
+
+  // hasPreferredResources will be false if all resources are selected
+  const maxResources = Object.keys(ClusterResource).length / 2 - 1;
```

Comment:
Hmm how does it work? Why are we dividing by 2? (I know that it was there previously, but I'm curious)

Reply:
Because ClusterResource is a numeric enum, not a  string enum, `Object.keys` and `Object.values` will return both keys and values. 

We could alternatively add a filter `.filter((v) => !isNaN(Number(v)))` and take the length.